chainstate/tx-verifier/src/transaction_verifier/input_output_policy/constraints_accumulator.rs

```diff
@@ -13,705 +13,584 @@
-        let inputs_utxos = vec![
-            Some(TxOutput::CreateStakePool(
-                pool_id,
-                Box::new(stake_pool_data),
-            )),
-            Some(TxOutput::Transfer(
-                OutputValue::Coin(Amount::from_atoms(transferred_atoms)),
-                Destination::AnyoneCanSpend,
-            )),
-            None,
-        ];
-
-        let outputs = vec![
-            TxOutput::LockThenTransfer(
-                OutputValue::Coin(Amount::from_atoms(staked_atoms + delegated_atoms)),
-                Destination::AnyoneCanSpend,
-                OutputTimeLock::ForBlockCount(
-                    required_decommission_maturity as u64 + required_spend_share_maturity as u64
-                        - 1,
-                ),
-            ),
-            TxOutput::Transfer(
-                OutputValue::Coin(Amount::from_atoms(transferred_atoms)),
-                Destination::AnyoneCanSpend,
-            ),
-        ];
-
-        let mut constraints_accumulator = ConstrainedValueAccumulator::new();
-        let result = constraints_accumulator.process_outputs(&chain_config, &outputs).unwrap_err();
-        assert_eq!(
-            result,
-            IOPolicyError::AttemptToPrintMoneyOrViolateTimelockConstraints
-        );
+fn decrease_or(
+    total_amounts: &mut BTreeMap<CoinOrTokenId, Amount>,
+    key: CoinOrTokenId,
+    amount: Amount,
+    error: IOPolicyError,
+) -> Result<(), IOPolicyError> {
+    if amount > Amount::ZERO {
+        match total_amounts.get_mut(&key) {
+            Some(value) => {
+                *value = (*value - amount).ok_or(error)?;
+            }
+            None => {
+                return Err(error);
+            }
+        }
+    }
+    Ok(())
+}
 
-        // valid case
-        let outputs = vec![
-            TxOutput::LockThenTransfer(
-                OutputValue::Coin(Amount::from_atoms(staked_atoms + delegated_atoms)),
-                Destination::AnyoneCanSpend,
-                OutputTimeLock::ForBlockCount(
-                    required_decommission_maturity as u64 + required_spend_share_maturity as u64,
-                ),
-            ),
-            TxOutput::Transfer(
-                OutputValue::Coin(Amount::from_atoms(transferred_atoms)),
-                Destination::AnyoneCanSpend,
-            ),
-        ];
-
-        let mut constraints_accumulator = ConstrainedValueAccumulator::new();
-
-        constraints_accumulator
-            .process_inputs(
-                &chain_config,
-                BlockHeight::new(1),
-                pledge_getter,
-                &inputs,
-                &inputs_utxos,
-            )
-            .unwrap();
-
-        constraints_accumulator.process_outputs(&chain_config, &outputs).unwrap();
-
-        assert_eq!(constraints_accumulator.consume().unwrap(), Amount::ZERO);
+fn merge_amount_maps<K: Ord>(
+    left: &mut BTreeMap<K, Amount>,
+    right: BTreeMap<K, Amount>,
+) -> Result<(), IOPolicyError> {
```

Comment:
What slightly confused me when reading the code that uses this is the naming. `merge_` sounds like a pure function but it actually modifies the first argument. Since it behaves somewhat like `std::ops::AddAssign`, I'd prefer this to be called `merge_assign_amount_maps`.

Alternatively, if you can afford to change this method to take ownership of the first argument, it may make sense to change the method to have the following signature (where `AmountMap` = `BTreeMap<K, Amount>`):
```rust
fn merge(left: AmountMap, right: AmountMap) -> Result<AmountMap, IOPolicyError>;
```

That is, move the map into the merge function and return the modified one as a result. It has some advantages:
* More mathematical/functional interface avoiding implicit mutation of one of the arguments
* No danger of using the `left` map after an error.
* Could be implemented as an impl of `std::ops::Add` to get the `+` operator.
* The updates could still be performed in place.

The drawback is that it now requires passing the map into the function rather than just a reference.

Reply:
Agree, reworked all the methods